### PR TITLE
chore: Move to monorepo ❗ ❗ ❗ 

### PR DESCRIPTION
**Pull request by benlesh**
- Reorganized almost all files in the project:
  - `packages/rxjs`: The new home for the main `rxjs` package.
  - `apps/rxjs.dev`: The new home for the docs app
  - `packages/*`: will be where other future packages live (i.e. `@rxjs/observable`)
  - We may move the decision tree project out from under the docs app to the `apps/*` directory at some point.
- "Fixed" the docs app tests: We've had an issue where the many of the tests were broken following an update to the docs up. I fixed this by simply deleting the broken tests (I know) after attempting to debug them and realizing it was a huge time suck. The principle here is that it's better to run _some_ tests for the docs app in CI than run _no_ tests in CI for the docs app.
- Updated `lodash` (and `@types/lodash`): This was done to fix problems with the TS@latest tests we run in CI. This is because the newer version of TypeScript fixed the types for WeakMap, and the older version of `@types/lodash` had them redefined differently. This is a small change.
- Updated a LOT of locations in a lot of build-related files. The changes are obviously to just point from `./` to `./packages/rxjs/` in some cases.
- Added a script to ensure the license file and code of conduct are duplicated to be under the `rxjs` package when we publish.
- **moved to yarn!!**: This is a big one. I _did not want to do this_, but I had no choice. Basically, `npm install` for workspaces will hoist all packages to the top level. When that's done, both `@types/mocha` (from rxjs) and `@types/jasmine` (from the docs app) are hoisted to the root `node_modules` folder. When that happens, TypeScript chokes on build because of duplicate type definitions for things like `describe`, `xit`, etc. Yarn features a nice `nohoist` setting, and we'll have to stick with Yarn until [npm decides how they're going to fix it](https://github.com/npm/rfcs/issues/287#issuecomment-1727960500)

### CODE REVIEW

**Comment by benlesh** [comment]
Note that the older actions "Node 16 build" and "ts@latest (18)" can be ignored for this PR.

image

**Comment by benlesh** [comment]
Note to @jakovljevic-mladen and @niklas-wortmann: To build or run the docs app now you'll want to use yarn workspaces: `yarn workspace rxjs.dev start` for example. Pretty straightforward, but I wanted to call it out.

After you initially pull this down, you'll want to delete your `node_modules`, then run `yarn install`, after that you're good to go.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -27,8 +27,8 @@ Contents
 ---
 
 - Related documents
-  - [Creating Operators](docs_app/content/guide/operators.md#creating-custom-operators)
-  - [Writing Marble Tests](docs_app/content/guide/testing/marble-testing.md)
+  - [Creating Operators](../../apps/rxjs.dev/content/guide/operators.md#creating-custom-operators)
```
Comment (jakovljevic-mladen) [inline CONTRIBUTING.md:30]: This isn't really working. As far as I understand, we now have `apps/rxjs.dev/` that contains (almost) all content of the old `docs_app` folder. So, I guess this should work:

```suggestion
  - [Creating Operators](apps/rxjs.dev/content/guide/operators.md#creating-custom-operators)
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -27,8 +27,8 @@ Contents
...
+  - [Writing Marble Tests](../../apps/rxjs.dev/content/guide/testing/marble-testing.md)
```
Comment (jakovljevic-mladen) [inline CONTRIBUTING.md:31]:
```suggestion
  - [Writing Marble Tests](apps/rxjs.dev/content/guide/testing/marble-testing.md)
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -122,15 +122,15 @@ from the main (upstream) repository:
 
 - The documentation is auto-generated directly from the source code.
 - In short: From the source code we generate JSON documents, describing each operator, function ... and render this JSON within an Angular application
-- The folder `docs_app` contains everything you need for building and developing the docs
-- The [Documentation README](docs_app/README.md) will support you
+- The folder `rxjs.dev` contains everything you need for building and developing the docs
```
Comment (jakovljevic-mladen) [inline CONTRIBUTING.md:125]: Maybe we should have it changed to this:

```suggestion
- The folder `apps/rxjs.dev` contains everything you need for building and developing the docs
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -122,15 +122,15 @@ from the main (upstream) repository:
...
+- The [Documentation README](../../apps/rxjs.dev/README.md) will support you
```
Comment (jakovljevic-mladen) [inline CONTRIBUTING.md:126]:
```suggestion
- The [Documentation README](apps/rxjs.dev/README.md) will support you
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -122,15 +122,15 @@ from the main (upstream) repository:
...
 - After a PR is merged to master the docs will be published to https://rxjs.dev/
 
 ## Unit Tests
 
 Unit tests are located under the [spec directory](/spec). Unit tests over synchronous operators and operations
 can be written in a standard [chai](https://www.chaijs.com/) style. Unit tests written against any
-asynchronous operator should be written in [Marble Test Style outlined in detail here](docs_app/content/guide/testing/marble-testing.md).
+asynchronous operator should be written in [Marble Test Style outlined in detail here](../../apps/rxjs.dev/content/guide/testing/marble-testing.md).
```
Comment (jakovljevic-mladen) [inline CONTRIBUTING.md:133]:
```suggestion
asynchronous operator should be written in [Marble Test Style outlined in detail here](apps/rxjs.dev/content/guide/testing/marble-testing.md).
```

File: README.md [hunk 6]
```diff
@@ -1,22 +1,22 @@
-# <img src="docs_app/src/assets/images/logos/Rx_Logo_S.png" alt="RxJS Logo" width="86" height="86"> RxJS: Reactive Extensions For JavaScript
+# <img src="apps/rxjs.dev/src/assets/images/logos/Rx_Logo_S.png" alt="RxJS Logo" width="86" height="86"> RxJS: Reactive Extensions For JavaScript
```
Comment (jakovljevic-mladen) [inline README.md:1]: Yeah, [it works](https://github.com/ReactiveX/rxjs/tree/b4d44efd123a845ef3770975175f6e3eb53162f3) in this case, so I guess this path will work in `CONTRIBUTING.md` file as well.

File: apps/rxjs.dev/src/app/app.component.ts [hunk 7]
```diff
@@ -231,15 +231,22 @@ export class AppComponent implements OnInit {
...
-        (version) => version.title === this.deployment.mode || version.title === `v${versionInfo.major}`
-      )!;
-      this.currentDocVersion.title += ` (v${versionInfo.raw})`;
+    ).subscribe({
+      next: ([versionInfo, versions]) => {
+        this.docVersions = [...versions];
+
+        // Find the current version - either title matches the current deployment mode
+        // or its title matches the major version of the current version info
+        console.log(this.docVersions);
```
Comment (jakovljevic-mladen) [inline apps/rxjs.dev/src/app/app.component.ts:240]: Please remove this line.

File: apps/rxjs.dev/src/app/app.component.ts [hunk 8]
```diff
@@ -231,15 +231,22 @@ export class AppComponent implements OnInit {
...
+
+        this.currentDocVersion = this.docVersions.find(
+          (version) => version.title === this.deployment.mode || version.title === `v${versionInfo.major}`
+        )!;
+        this.currentDocVersion.title += ` (v${versionInfo.raw})`;
+      },
+      error: (err) => {
```
Comment (jakovljevic-mladen) [inline apps/rxjs.dev/src/app/app.component.ts:247]: And this one.

File: apps/rxjs.dev/src/app/custom-elements/operator-decision-tree/README.md [hunk 9]
```diff
@@ -1,8 +1,9 @@
...
-The JSON is generated via the `docs_app/tools/decision-tree-generator`.
+The JSON is generated via the `apps/rxjs.dev/tools/decision-tree-generator`.
 
 # TODO
+
 - Consider placing the widget on the home page - or a link on the home page, “Decision Tree”
 - Manual focus calls when navigating the tree (example: after making a selection, focus on the current sentence)
-- Drop jasmine-marbles for just the TestScheduler
\ No newline at end of file
+- Drop jasmine-marbles for just the TestScheduler
```
Comment (jakovljevic-mladen) [inline apps/rxjs.dev/src/app/custom-elements/operator-decision-tree/README.md:9]: I think I dropped this in #7238. Or at least I wanted to. Whatever it is, I will have to think how to fix that PR to work with these changes.

File: apps/rxjs.dev/tools/transforms/angular-base-package/index.js [hunk 10]
```diff
@@ -37,7 +37,7 @@ module.exports = new Package('angular-base', [
 
   // overrides base packageInfo and returns the one for the 'angular/angular' repo.
   .factory('packageInfo', function () {
-    return require(path.resolve(PROJECT_ROOT, 'package.json'));
+    return require(path.resolve(PROJECT_ROOT, './packages/rxjs/package.json'));
```
Comment (jakovljevic-mladen) [inline apps/rxjs.dev/tools/transforms/angular-base-package/index.js:40]: I think the `PROJECT_ROOT` path should be fixed, not the other way around. Suggestion (after `PROJECT_ROOT` path is fixed):

```suggestion
    return require(path.resolve(PROJECT_ROOT, 'package.json'));
```

File: apps/rxjs.dev/tools/transforms/config.js [hunk 11]
```diff
@@ -1,25 +1,36 @@
 const { resolve } = require('path');
 const { readdirSync } = require('fs');
 
-const PROJECT_ROOT = resolve(__dirname, '../../..');
-const AIO_PATH = resolve(PROJECT_ROOT, 'docs_app');
+const PROJECT_ROOT = resolve(__dirname, '../../../..');
```
Comment (jakovljevic-mladen) [inline apps/rxjs.dev/tools/transforms/config.js:4]:
```suggestion
const PROJECT_ROOT = resolve(__dirname, '../../../../packages/rxjs');
```

File: packages/rxjs/README.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
+# <img src="apps/rxjs.dev/src/assets/images/logos/Rx_Logo_S.png" alt="RxJS Logo" width="86" height="86"> RxJS: Reactive Extensions For JavaScript
+
+![CI](https://github.com/reactivex/rxjs/workflows/CI/badge.svg)
+[![npm version](https://badge.fury.io/js/rxjs.svg)](http://badge.fury.io/js/rxjs)
+[![Join the chat at https://gitter.im/Reactive-Extensions/RxJS](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/Reactive-Extensions/RxJS?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
+
+# RxJS 8
+
+### FOR 7.X PLEASE GO TO [THE 7.x BRANCH](https://github.com/ReactiveX/rxjs/tree/7.x)
```
Comment (jakovljevic-mladen) [inline packages/rxjs/README.md:9]: If we're gonna have this file here, I don't think we have to mention V7 here.

Nor any mention of the RxJS versions.

File: packages/rxjs/README.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+range(1, 200)
+  .pipe(
+    filter((x) => x % 2 === 1),
+    map((x) => x + x)
+  )
+  .subscribe((x) => console.log(x));
+```
+
+If you're using RxJS version below 7.2, you can pull in any operator you need from one spot, under `'rxjs/operators'`.
```
Comment (jakovljevic-mladen) [inline packages/rxjs/README.md:54]: I guess this could be removed as well.

File: packages/rxjs/README.md [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+## Building/Testing
+
+- `yarn compile` build everything
+- `npm test` run tests
+- `yarn dtslint` run dtslint tests
+
+## Adding documentation
+
+We appreciate all contributions to the documentation of any type. All of the information needed to get the docs app up and running locally as well as how to contribute can be found in the [documentation directory](./apps/rxjs.dev).
```
Comment (jakovljevic-mladen) [inline packages/rxjs/README.md:84]:
```suggestion
We appreciate all contributions to the documentation of any type. All of the information needed to get the docs app up and running locally as well as how to contribute can be found in the [documentation directory](apps/rxjs.dev).
```

File: packages/rxjs/README.md (hunk 12 above)
Comment (benlesh) [reply packages/rxjs/README.md:9]: That can be done in a follow up.

File: apps/rxjs.dev/src/app/custom-elements/operator-decision-tree/README.md (hunk 9 above)
Comment (benlesh) [reply apps/rxjs.dev/src/app/custom-elements/operator-decision-tree/README.md:9]: There shouldn't be any real changes in this PR, it's just moving files and updating build scripts.

**Comment by benlesh** [comment]
@jakovljevic-mladen I had to revert both of your suggestions for resolving directory/file locations that I committed above. They broke the docs app build. If you think those are incorrect, I suggest we fix it in a follow up.

Otherwise, I think this is good to go.

**Comment by benlesh** [comment]
I'm going to go ahead and merge this so we can move on. @jakovljevic-mladen you can double-check the docs app to make sure everything is fine, but I ran it locally and ran through a few user flows just to see if it was okay, and everything checked out fine.

**Pull request merged**